source/osh/ui/view/spectrogram/SpectrogramView.js

```diff
@@ -0,0 +1,231 @@
+        this.maxDataPoints = 1000;
+        this.isSim = false;
+        this.tickArray = [];
+        for (let i = 0; i <= this.numBands; i++) {
+            let interval = (this.maxFreq - this.minFreq) / this.numBands;
+            let tickVal = i * interval;
+            this.tickArray.push(tickVal);
+        }
+
+
+        // Layout Vars
+        let thisDiv = document.getElementById(this.divId);
+        thisDiv.style.height = thisDiv.parentElement.getBoundingClientRect().height + 'px';
+        // this.width = thisDiv.getBoundingClientRect().width;
+        // this.height = thisDiv.getBoundingClientRect().height;
+        this.width = 500;
+        this.height = 350;
+        this.xOffset = (0.10 * this.width);
+        this.yOffset = (0.90 * this.height);
+        this.yOffsetBottom = 10;
+
+        // Draw Spectrogram
+        // The div that will be the root of the D3JS Chart
+        this.container = d3.select('#' + this.divId);
+        this.width = parseInt(this.container.style('width'), 10);
+        this.height = parseInt(this.container.style('height'), 10);
+        this.aspect = this.width / this.height;
+
+        console.log(this.width, this.height, this.aspect);
+
+        this.svg = this.container.append('svg')
+            .classed('spec-svg', true)
+            .attr("viewBox", `0 0 ${this.width} ${this.height}`)
+            .attr('preserveAspectRatio', 'xMinYMid')
+            .call(this.makeResponsive);
+
+        console.log(this.svg);
+
+        // Setup Scales
+        this.initXScale = d3.scaleTime()
+            .domain([this.minDate, this.maxDate])
+            .range([this.xOffset, this.width - 100]);
+        this.initXAxis = d3.axisBottom(this.initXScale);
+
+        this.initYScale = d3.scaleLinear()
+            .domain([this.minFreq, this.maxFreq])
+            .range([this.yOffset, this.yOffsetBottom]);    // TODO: name {bottomOffset} something better
+        // this.initYAxis = d3.axisLeft(this.initYScale).ticks(this.numBands).tickFormat((d, i) => d + this.freqUnit);
+        this.initYAxis = d3.axisLeft(this.initYScale).tickValues(this.tickArray).tickFormat((d, i) => d + this.freqUnit);
+
+        this.initZScale = d3.scaleSequential(d3.interpolateOrRd).domain(this.powerRange);
+
+
+        // Add Axes to Chart
+        this.xBar = this.svg.append('g')
+            .attr('class', ' x axis')
+            .attr('transform', 'translate(' + (0) + ',' + (this.yOffset) + ')')
+            .call(this.initXAxis);
+        this.yBar = this.svg.append('g')
+            .attr('class', 'y axis')
+            .attr('transform', 'translate(' + (this.xOffset) + ',' + (0) + ')')
+            .call(this.initYAxis);
+
+        if (options.hasOwnProperty('isSim')) {
+            this.isSim = options.isSim;
+        }
+
+        if (this.isSim) {
+            console.log("Simulating Spectrogram")
+            setInterval(() => {
+                this.spectrogramData = this.spectrogramData.concat(this.createDataEntries())
```

Comment:
You can use the spread operator which is much more efficient to concat array. I 've made a jsperf:
https://jsperf.com/concat-arrays-es6

Reply:
Definitely a good catch there. Fixed in the latest along with removing some extraneous comments